src/main/scala/akka/persistence/cassandra/query/journal/EventsByPersistenceIdPublisher.scala

```diff
@@ -0,0 +1,91 @@
+package akka.persistence.cassandra.query.journal
+
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+import com.datastax.driver.core.Session
+
+import akka.actor.Props
+import akka.persistence.PersistentRepr
+import akka.persistence.cassandra.EventsByPersistenceIdIterator
+import akka.persistence.query.EventEnvelope
+import akka.serialization.SerializationExtension
+
+private[journal] object EventsByPersistenceIdPublisher {
+  def props(
+      persistenceId: String,
+      fromSeqNr: Long,
+      toSeqNr: Long,
+      refreshInterval: Option[FiniteDuration],
+      maxBufSize: Long,
+      session: Session,
+      config: CassandraReadJournalConfig): Props =
+    Props(
+      new EventsByPersistenceIdPublisher(
+        persistenceId,
+        fromSeqNr,
+        toSeqNr,
+        refreshInterval,
+        maxBufSize,
+        session,
+        config))
+}
+
+// TODO: Decouple database.
+// TODO: Query index tables instead.
+// TODO: Generic message iterator to handle different tables etc.
+private[journal] class EventsByPersistenceIdPublisher(
+    persistenceId: String,
+    fromSeqNr: Long,
+    toSeqNr: Long,
+    refreshInterval: Option[FiniteDuration],
+    maxBufSize: Long,
+    session: Session,
+    override val config: CassandraReadJournalConfig)
+  extends QueryActorPublisher[EventEnvelope, Long](refreshInterval, maxBufSize)
+  with CassandraReadStatements {
+
+  import config._
+
+  private[this] val serialization = SerializationExtension(context.system)
+
+  private[this] val step = 50l
+
+  override protected def query(state: Long, max: Long): Future[Vector[EventEnvelope]] = {
+    implicit val ec = context.dispatcher
+
+    // TODO: Async?
+    Future {
```

Comment:
As mentioned above the database queries are expected to be asynchronous. Causality here can be achieved when updating buffer. This class is only concerned about the buffer and state update logic, nothing else. Database and stream publisher are abstracted elsewhere.